Bonus guide: Homer

VajraOfIndra opened the pull request:
#### What

A web service dashboard

### Why/How

See discussions in closed PR #968 and in Issue # 849

#### Scope

- [ ] significant change to core configuration
- [x] independent bonus guide
- [ ] simple bug fix

RooSoft commented on guide/bonus/raspberry-pi/homer.md:
````diff
@@ -0,0 +1,610 @@
+* Create a symlink to the logos folder and change its ownership
+
+  ```sh
+  $ sudo ln -s /data/homer/tools /var/www/homer/assets
````
The tools folder is being used by Homer, such as in this URL: `https://raspibolt:4091/assets/tools/raspibolt3.png`

Means, I think the command above should be changed to:

```bash
sudo ln -s /data/homer /var/www/homer/assets
```

... so that the tools folder ends up existing under the assets folder.

Otherwise, links will need to be changed in the front end to remove the tools folder.

VajraOfIndra replied:
> ... so that the tools folder ends up existing under the assets folder.

I believe that's what the existing symlink provides. It create a "tools" symlink in `/var/www/homer/assets` that redirects to the tools folder in `/data/homer` so that when Homer looks for `/assets/tools` it sees all  the images inside `/data/homer/tools`.

> Means, I think the command above should be changed to:

> `sudo ln -s /data/homer /var/www/homer/assets`

This creates the following symlink inside /assets: `homer -> /data/homer` Afaik, the tools folder would then be missing and Homer wouldn't be able to display any images...

But I might be misunderstanding your point.

RooSoft replied:
Problem is that at that point, I don't think the `/var/www/homer/assets` folder exists... so the `/data/homer/tools` folder ends up being linked to by `/var/www/homer/assets` directly. When the website tries to reach files in the `assets/tools` folder, it issues some 404 because that folder doesn't exist.

Try it by deleting the `/var/www/homer/assets` folder and running the above command. The contents of the assets folders should be what tools is supposed to be and thus the tools folder will be absent.

VajraOfIndra replied:
> Problem is that at that point, I don't think the `/var/www/homer/assets` folder exists..

Hum... it should exist afaik. It is being created due to the following command (which is done a bit before, like 2 commands before):
`sudo rsync -av --delete /home/homer/homer/dist/ /var/www/homer/`

The rsync dump the content of the dist folder into /var/www/homer, and it includes the tools folder...

--- 

But let me delete Homer and follow the guide again, I'll see if I can reproduce your issue.
I'll get back to you once I tested this again :+1:

VajraOfIndra replied:
I deleted Homer frm my node and rebuilt it from scratch.

I think the issue was when creating the symlink for the "tools" folder. There is a need to first delete the existing folder which contains sample pics. I added a rm line to do this just before creating the symlink.

I also shuffled a few commands to simplify the guide.

I went thourgh the whole guide and can display the homepage well with all the logos in my browser.

PTAL

RooSoft replied:
Done my homework and it seems to work...

Figuring out that the `ln` command acts differently if there is a trailing slash or not.. with a slash, it copies the contents of the assets folder. When it's not there, makes a link to the same folder instead and that's what's wanted and documented.

Well done!

VajraOfIndra replied:
Ok thanks, I'll try to keep in mind that trailing slash effect on symlink creation 👍

RooSoft commented on guide/bonus/raspberry-pi/homer.md:
````diff
@@ -0,0 +1,610 @@
+
+  ```sh
+  $ sudo ln -s /data/homer/tools /var/www/homer/assets
+  $ sudo chown -h www-data:www-data /var/www/homer/assets/tools
````
Here, if we do as above, the assets should be chown, with a recursive flag so that the tools folder gets affected as well.

```bash
sudo chown -R -h www-data:www-data /var/www/homer/assets
```

VajraOfIndra replied:
I now place the chown command at the very end just before the first launch. Instead of chowning several times, it is now done only once which simplifies the guide.

RooSoft commented on guide/bonus/raspberry-pi/homer.md:
```diff
@@ -0,0 +1,610 @@
+
+
+  ################################
+  # Services                     #
```
I guess this section could be a whole document in itself, as services could differ wildly between instances of Raspibolt.

For instance, it took me a while to understand that the White paper link was referring to a section I personally haven't applied on mine, hence the resulting 404.

VajraOfIndra replied:
The idea was to include only web services that have been created if somone follows the core guide, i.e. the BTC RPC Explorer and RTL. The white paper is actually also the BTC RPC Explorer (just a specific page/URL from it) so it should work for anyone who has installed the Explorer.

Maybe, I could add a warning at the start that the guide assumes that the user followed the main guide ans had installed the Explorer and RTL.

VajraOfIndra replied:
We could also remove the white paper download as it's not a web service per se but a special link from the Explorer. 

I just thought it would add a bit of content to the dashboard and also increase the visibility of this little known feature of the blockchain.

Shall we remove it?

VajraOfIndra replied:
Ok, so I added a warning in the introduction: 
![image](https://user-images.githubusercontent.com/86242283/163953384-557a1b62-ad05-438b-8417-00d1d22d40fd.png)

VajraOfIndra replied:
Hum, I think I understand your point now.

I could maybe write a "Configuration" section, where we give the default config file (as it is now), followed by a series of short explanations on how to edit Homer and make it more personal: how to remove/add a web service, how to use logo vs icon, how to change the API-based message etc. Either in the middle of the guide (under a "Configuration" section, or towards the end, after the default install, under a "Personalize your dashboard" chapter.

Wdyt?

VajraOfIndra replied:
I deleted Homer frm my node and rebuilt it from scratch.

I think the issue was when creating the symlink for the "tools" folder. There is a need to first delete the existing folder which contains sample pics. I added a rm line to do this just before creating the symlink.

I also shuffled a few commands to simplify the guide.

I went thourgh the whole guide and can display the homepage well with all the logos in my browser.

PTAL

VajraOfIndra replied:
I created a brand new chapter that introduce users on how to personalize their dashboards and add new web service buttons. I think the guide is quite complete now :)

RooSoft reviewed (changes requested):
Very impressive document, I got to say... applied it as is from start to finish with very little problems aside from the tools folder.

Kudos to @VajraOfIndra!

VajraOfIndra commented:
Thanks @RooSoft for taking the time to review this PR!

I replied to your reviews, PTAL.

(I also added a few missing hyperlinks and updated the top illustration to match the guide in its present-day form)

RooSoft reviewed (changes requested):
The `assets` and `tools` folders documentation still needs some correction

RooSoft commented:
In my review, I think I forgot about that 404 the page hits in the background... no idea what it is.

`https://raspibolt:4091/assets/manifest.json`

kristapsk commented on guide/bonus/raspberry-pi/homer.md:
````diff
@@ -0,0 +1,612 @@
+  ```sh
+  $ cd homer
+  $ git fetch
+  $ git describe --tags --abbrev=0
````
```suggestion
  $ git tag | sort --version-sort | tail -n 1
```

VajraOfIndra replied:
Thanks :+1:

VajraOfIndra commented:
> The `assets` and `tools` folders documentation still needs some correction



> In my review, I think I forgot about that 404 the page hits in the background... no idea what it is.
> 
> `https://raspibolt:4091/assets/manifest.json`

Where do you see this error? in the browser or in some logs?

The` manifest.json` is part of the `assets` directory. Did you delete the file or folder maybe?
I can't replicate this issue. I can see the file in my asset folder:
![image](https://user-images.githubusercontent.com/86242283/164912228-0127ec2e-8a45-42f9-a371-c900b38f9db6.png)

RooSoft commented:
> > The `assets` and `tools` folders documentation still needs some correction
> 
> > In my review, I think I forgot about that 404 the page hits in the background... no idea what it is.
> > `https://raspibolt:4091/assets/manifest.json`
> 
> Where do you see this error? in the browser or in some logs?
> 
> The` manifest.json` is part of the `assets` directory. Did you delete the file or folder maybe? I can't replicate this issue. I can see the file in my asset folder: ![image](https://user-images.githubusercontent.com/86242283/164912228-0127ec2e-8a45-42f9-a371-c900b38f9db6.png)

In the browser

RooSoft reviewed (approved):
Incredibly well crafted piece of documentation right from the get go.

VajraOfIndra commented:
Thanks a lot @RooSoft and @kristapsk for your reviews! 👍

It's now in master and live 🎉
